unstructured/partition/docx.py

```diff
@@ -134,305 +151,625 @@ def partition_docx(
-    if filename is not None:
-        if not filename.startswith("/tmp"):
-            last_modification_date = get_last_modified_date(filename)
-
-        document = docx.Document(filename)
-    elif file is not None:
-        last_modification_date = get_last_modified_date_from_file(file)
-
-        document = docx.Document(
-            spooled_to_bytes_io_if_needed(
-                cast(Union[BinaryIO, SpooledTemporaryFile], file),
-            ),
+    return list(
+        _DocxPartitioner.iter_document_elements(
+            filename,
+            file,
+            metadata_filename,
+            include_page_breaks,
+            metadata_last_modified,
         )
+    )
 
-    elements: List[Element] = []
-    table_index = 0
-
-    headers_and_footers = _get_headers_and_footers(document, metadata_filename)
-    if len(headers_and_footers) > 0:
-        elements.extend(headers_and_footers[0][0])
-
-    document_contains_pagebreaks = _element_contains_pagebreak(document._element)
-    page_number = 1 if document_contains_pagebreaks else None
-    section = 0
-    is_list = False
-    for element_item in document.element.body:
-        if element_item.tag.endswith("tbl"):
-            table = document.tables[table_index]
-            emphasized_texts = _get_emphasized_texts_from_table(table)
-            emphasized_text_contents, emphasized_text_tags = _extract_contents_and_tags(
-                emphasized_texts,
-            )
-            html_table = convert_ms_office_table_to_text(table, as_html=True)
-            text_table = convert_ms_office_table_to_text(table, as_html=False)
-            element = Table(text_table)
-            if element is not None:
-                element.metadata = ElementMetadata(
-                    text_as_html=html_table,
-                    filename=metadata_filename,
-                    page_number=page_number,
-                    last_modified=metadata_last_modified or last_modification_date,
-                    emphasized_text_contents=emphasized_text_contents,
-                    emphasized_text_tags=emphasized_text_tags,
-                )
-                elements.append(element)
-            table_index += 1
-        elif element_item.tag.endswith("p"):
-            if "<w:numPr>" in element_item.xml:
-                is_list = True
-            paragraph = docx.text.paragraph.Paragraph(element_item, document)
-            emphasized_texts = _get_emphasized_texts_from_paragraph(paragraph)
-            emphasized_text_contents, emphasized_text_tags = _extract_contents_and_tags(
-                emphasized_texts,
+
+class _DocxPartitioner:
+    """Provides `.partition()` for MS-Word 2007+ (.docx) files."""
+
+    # TODO: I think we can do better on metadata.filename. Should that only be populated when a
+    #       `metadata_filename` argument was provided to `partition_docx()`? What about when not but
+    #       we do get a `filename` arg or a `file` arg that has a `.name` attribute?
```

Comment:
I think it is ok to have it the way it is (e.g. `None` when no argument was provided). One way to think about this is maybe we have another field to record the filename if provided, and fall back to `.name` when not found. But I'll take more opinions on this too!

Reply:
@Klaijan Present behavior for `partition_docx(filename="some_file.docx")` is to record `None` for `metadata.filename`. I'm thinking `metadata_filename` should take precedence, but isn't `filename` better than `None` for the case where they're just partitioning a file on a filesystem?